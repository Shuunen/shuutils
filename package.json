{
  "$schema": "https://json.schemastore.org/package",
  "author": "Romain Racamier-Lafon <romain.racamier@gmail.com>",
  "bin": {
    "unique-mark": "./dist/unique-mark.js"
  },
  "bugs": "https://github.com/Shuunen/shuutils/issues",
  "dependencies": {
    "tiny-glob": "0",
    "uvu": "0"
  },
  "description": "utils collection",
  "devDependencies": {
    "@happy-dom/global-registrator": "8",
    "@types/node": "18",
    "@vitest/coverage-c8": "0.29",
    "dependency-cruiser": "12",
    "esbuild": "0.17",
    "eslint": "8.36",
    "eslint-config-hardcore": "31",
    "eslint-plugin-jsdoc": "40",
    "repo-check": "1.30",
    "typescript": "4",
    "vitest": "0.29",
    "watchlist": "0"
  },
  "files": [
    "dist"
  ],
  "homepage": "https://github.com/Shuunen/shuutils#readme",
  "keywords": [
    "shuunen",
    "utils",
    "collection",
    "pure",
    "javascript",
    "typescript"
  ],
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "name": "shuutils",
  "private": false,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Shuunen/shuutils.git"
  },
  "scripts": {
    "build": "rm dist -rf && tsc --emitDeclarationOnly && node bin/build && echo build success",
    "build:watch": "watchlist bin src -eager -- npm run build",
    "check": "repo-check",
    "ci": "npm run build && npm run lint && npm run check && npm run test && echo ci success",
    "dev": "npm run test:watch",
    "lint": "tsc --noEmit && eslint --cache --fix --ignore-path .gitignore --ext .js,.ts . && depcruise src --config && echo lint success",
    "lint:watch": "watchlist src tests -eager -- npm run lint",
    "postversion": "git push && git push --tags && npm publish",
    "prepack": "npm run build",
    "preversion": "npm run ci",
    "start": "npm run dev",
    "test": "npm run test:watch -- --run && echo test success",
    "test:watch": "vitest --coverage"
  },
  "type": "module",
  "types": "dist/src/index.d.ts",
  "version": "7.2.0"
}