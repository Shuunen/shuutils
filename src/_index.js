'use strict'

import { getTimestamp, getTimestampMs } from './timestamps'

module.exports = { getTimestamp, getTimestampMs }