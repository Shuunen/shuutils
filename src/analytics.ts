/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types */
import { sleep } from './functions'
import { Logger } from './logger'

type AnalyticsInstance = {
  identify: (userId: string, additionalData?: Record<string, unknown>) => Promise<void>
  page: () => Promise<void>
  track: (event: string, additionalData?: Record<string, unknown>) => Promise<void>
}

/**
 * Generic analytics wrapper, can be used with any analytics third party library
 * @example const analytics = new Analytics()
 * @example analytics.setupInMemory('my-app')
 * @example analytics.track('invoice-download')
 * @example analytics.track('click', { category: '...', label: '...', ... })
 */
// eslint-disable-next-line no-restricted-syntax
export class Analytics {
  private $app = ''
  private $instance: AnalyticsInstance | undefined = undefined
  private readonly $logger = new Logger()
  public readonly pile: string[] = []

  /**
   * Identify a user
   * @param userId - The user, like '123456', 'super-admin', etc.
   * @param additionalData - Additional data to identify the user, like `{ email: '...', age: 35, ... }`
   */
  public identify(userId: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : identify failed, analytics not setup')
    this.$logger.info(`analytics : identify user "${userId}" on app "${this.$app}"`)
    void this.$instance.identify(userId, additionalData)
  }

  /**
   * Track a page view
   */
  public page() {
    if (this.$instance === undefined) throw new Error('analytics : track page failed, analytics not setup')
    this.$logger.debug(`analytics : track page for app "${this.$app}"`)
    void this.$instance.page()
  }

  /**
   * Setup analytics with in memory tracking, useful for tests
   * @param app - The name of the app, like 'mlm-react-front', 'cosmos-front', etc.
   */
  public setupInMemory(app: string) {
    this.$app = app
    // eslint-disable-next-line jsdoc/require-jsdoc
    const addToPile = async (item: string, data?: Record<string, unknown>) => {
      this.$logger.debug(`analytics : add "${item}" to pile`, data)
      await sleep(1)
      this.pile.push(item)
      this.$logger.debug(`analytics : pile for app "${app}"`, this.pile)
    }
    this.$instance = {
      // eslint-disable-next-line jsdoc/require-jsdoc
      identify: async (userId: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`identify:${userId}`, additionalData)
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      page: async () => {
        await addToPile('page')
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      track: async (event: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`track:${event}`, additionalData)
      },
    }
    this.$logger.info(`analytics : setup app "${app}" with in memory tracking`)
  }

  /**
   * Track an event
   * @param event - The event name like 'click', 'download-invoice', etc.
   * @param additionalData - Additional data to track the event, like `{ category: '...', label: '...', ... }`
   */
  public track(event: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : track event failed, analytics not setup')
    this.$logger.debug(`analytics : track event "${event}" for app "${this.$app}"`)
    void this.$instance.track(event, additionalData)
  }
}
