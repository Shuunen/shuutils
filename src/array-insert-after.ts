import { nbAfter, nbNoIndex } from './constants'
import { clone } from './object-clone'

/**
 * Add a value to an array if it doesn't already exist
 * @param array the array to add the value to
 * @param item the value to add
 * @param value the value to add
 * @returns the array with the value added (if it didn't already exist)
 */
export function insertValueAfterItem<T> (array: T[], item: T, value: T) {
  const index = array.indexOf(item)
  if (index === nbNoIndex) return array
  const arrayCopy = clone(array)
  const start = Math.min(index + nbAfter, arrayCopy.length)
  arrayCopy.splice(start, 0, value)
  return arrayCopy
}
