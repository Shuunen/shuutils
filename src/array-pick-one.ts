/**
 * Return an item from an array
 * @param items like : ["great", "place", "pine"]
 * @returns item like : "pine"
 */
export function pickOne<Type>(items: readonly Type[]) {
  if (items.length === 0) throw new Error('Array is empty')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return items[Math.floor(Math.random() * items.length)] as Type
}
