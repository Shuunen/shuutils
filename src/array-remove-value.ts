import { clone } from './object-clone'

/**
 * Remove a value from an array
 * @param array the array to remove the value from
 * @param value the value to remove
 * @returns a copy of the array without the value
 */
export function removeValueFromArray<Type>(array: readonly Type[], value: Type) {
  if (!array.includes(value)) return array
  const arrayCopy = clone(array)
  arrayCopy.splice(arrayCopy.indexOf(value), 1)
  return arrayCopy
}
