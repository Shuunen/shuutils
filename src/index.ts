export * from './array-insert-after'
export * from './array-pick-one'
export * from './array-remove-value'
export * from './array-shuffle'
export * from './array-unique'
export * from './browser-clipboard'
export * from './browser-fill-like-human'
export * from './browser-inject-styles'
export * from './browser-page-change'
export * from './browser-scout'
export * from './browser-toast'
export * from './colors'
export * from './constants'
export * from './date-relative'
export * from './dates'
export * from './debounce'
export * from './dom'
export * from './environment'
export * from './events'
export * from './expect-type'
export * from './functions'
export * from './logger'
export * from './memoize'
export * from './number-random'
export * from './number-ranged-score'
export * from './number-round'
export * from './object-cap-nesting'
export * from './object-clone'
export * from './object-equal'
export * from './object-flatten'
export * from './object-serializer'
export * from './object-sort'
export * from './objects'
export * from './state'
export * from './storage'
export * from './string-remove-accents'
export * from './strings'
export * from './testing-align-for-snap'
export * from './throttle'
export * from './timestamps'
export * from './types'
