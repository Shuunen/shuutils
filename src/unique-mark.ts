import { execSync } from 'node:child_process'
import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import path from 'node:path'
import glob from 'tiny-glob'
import { formatDate } from './dates.js'
import { isTestEnvironment } from './environment.js'
import { Logger } from './logger.js'
import { injectMark, parseJson } from './strings.js'
import type { PackageJson } from './types'

/**
 * Load the package.json file data
 * @param location the location of the package.json file
 * @returns the package.json version
 */
export function getPackageJsonVersion(location = path.join(process.cwd(), 'package.json')) {
  if (!existsSync(location)) throw new Error(`package.json was not found in ${location}, aborting.`)
  const content = readFileSync(location, 'utf8')
  const { error, value } = parseJson<PackageJson>(content)
  if (error) throw new Error(`package.json in ${location} is not a valid JSON, aborting.`)
  return value.version
}

/**
 * Get the files to inject the mark in
 * @param target the glob to get the files from, like "public/index.html" or "public/*.js"
 * @returns the files to inject the mark in
 */
// eslint-disable-next-line @typescript-eslint/no-magic-numbers
export async function getTargetFiles(target = process.argv[2] ?? '') {
  if (target === '') throw new Error('no target specified, aborting.')
  // if ends with something like *.js
  const extension = /\.(?<ext>[a-z]+)$/u.exec(target)?.groups?.ext ?? ''
  if (extension !== '') throw new Error(`provided : "${target}", you need to use *.{${extension}} to capture all files with that extension (limitation of tiny-glob)`)
  const files = await glob(target)
  if (files.length === 0) throw new Error(`no file found for target "${target}", aborting.`)
  return files
}

/**
 * Generate the mark to inject
 * @param root0 the options
 * @param root0.commit the commit hash to use, if empty, will use the last git commit hash
 * @param root0.date the date to use, if empty, will use the current date
 * @param root0.version the version to use, if empty, will use the version from package.json
 * @returns the mark to inject, like "4.2.0 - 123abc45 - 01/01/2021 12:00:00"
 */
export function generateMark({ commit = '', date = formatDate(new Date(), 'dd/MM/yyyy HH:mm:ss'), version = '' }: Readonly<{ commit?: string; date?: string; version?: string }>) {
  let finalCommit = commit
  /* c8 ignore next */
  if (commit === '') finalCommit = execSync('git rev-parse --short HEAD', { cwd: process.cwd() }).toString().trim()
  return `${version} - ${finalCommit} - ${date}`
}

/**
 * Inject the mark in the files
 * @param root0 the options
 * @param root0.placeholder the placeholder to use to inject the mark
 * @param root0.mark the mark to inject
 * @param root0.files the files to inject the mark in
 * @param root0.isReadOnly if true, will not write the files to disk (useful for unit tests)
 * @returns the total amount of mark injection in the targeted files
 */
// eslint-disable-next-line max-statements, complexity
export function injectMarkInFiles({
  files = [],
  isReadOnly = false,
  mark = 'no-mark',
  placeholder = 'unique-mark',
}: Readonly<{ files?: readonly string[]; isReadOnly?: boolean; mark?: string; placeholder?: string }>) {
  let totalInjections = 0
  const logs: string[] = []
  const markRegex = new RegExp(mark, 'gu')
  for (const file of files) {
    const content = readFileSync(file, 'utf8')
    if (!content.includes(placeholder) && files.length === 1)
      throw new Error(
        `could not find a place to inject in ${file}, aborting.\n\nPlease use one or more of these placeholders :  <span id="${placeholder}"></span>  <meta name="${placeholder}" content="">  __${placeholder}__`,
      )
    const updatedContent = injectMark(content, placeholder, mark)
    const times = updatedContent.match(markRegex)?.length /* c8 ignore next */ ?? 0
    /* c8 ignore next 2 */
    if (!isReadOnly) writeFileSync(file, updatedContent)
    logs.push(`injected in ${file} : ${times} time${times > 1 ? 's' : ''}`)
    totalInjections += times
  }
  return { logs, totalInjections }
}

/* c8 ignore start */
/**
 * Main function
 */
async function init() {
  const logger = new Logger()
  logger.debug('starting...')
  const version = getPackageJsonVersion()
  const files = await getTargetFiles()
  logger.debug(`found ${files.length} file${files.length > 1 ? 's' : ''} to inject mark :`, files.join(', '))
  const mark = generateMark({ version })
  logger.debug('generated mark', mark)
  const { logs, totalInjections } = injectMarkInFiles({ files, mark })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-type-assertion
  for (const line of logs) logger.info(...(line.split(':') as [string, string]))
  if (totalInjections === 0) logger.info('files found but no mark found for injection')
  else logger.success('total injections :', String(totalInjections))
}

// eslint-disable-next-line unicorn/prefer-top-level-await
if (!isTestEnvironment()) void init()
