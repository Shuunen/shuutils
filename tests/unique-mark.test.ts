import { expect, it } from 'vitest'
import { version } from '../package.json'
import { generateMark, getPackageJsonVersion, getTargetFiles, injectMarkInFiles } from '../src/unique-mark'

it('getPackageJsonVersion A', () => {
  expect(getPackageJsonVersion()).toBe(version)
})

it('getPackageJsonVersion B pkg file not found', () => {
  expect(() => getPackageJsonVersion('non-existent.json')).toThrowErrorMatchingSnapshot()
})

it('getPackageJsonVersion C pkg file found but invalid json', () => {
  expect(() => getPackageJsonVersion('README.md')).toThrowErrorMatchingSnapshot()
})

it('getTargetFiles A list markdown files at root dir', async () => {
  expect(await getTargetFiles('*.md')).toMatchSnapshot()
})

it('getTargetFiles B list non existent files', async () => {
  await expect(async () => getTargetFiles()).rejects.toThrowErrorMatchingSnapshot()
})

const fakeMark = '9.7.8 - xyz - 07/05/2023 17:26:35'
it('generateMark A', () => {
  expect(generateMark({ commit: 'xyz', date: '07/05/2023 17:26:35', version: '9.7.8' })).toEqual(fakeMark)
})

it('injectMarkInFiles A successful', () => {
  expect(injectMarkInFiles({ files: ['src/strings.ts'], isReadOnly: true, mark: fakeMark, placeholder: 'placeholder' })).toMatchSnapshot()
})

it('injectMarkInFiles B fail to find placeholder', () => {
  expect(() => injectMarkInFiles({ files: ['src/strings.ts'], isReadOnly: true, mark: fakeMark, placeholder: 'nope' })).toThrowErrorMatchingSnapshot()
})
