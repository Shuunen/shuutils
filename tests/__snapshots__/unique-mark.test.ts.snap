// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPackageJsonVersion B pkg file not found 1`] = `"package.json was not found in non-existent.json, aborting."`;

exports[`getPackageJsonVersion C pkg file found but invalid json 1`] = `"package.json in README.md is not a valid JSON, aborting."`;

exports[`getTargetFiles A list markdown files at root dir 1`] = `
[
  "README.md",
]
`;

exports[`getTargetFiles B list non existent files 1`] = `"no file found for target \\"public/index.html\\", aborting."`;

exports[`injectMarkInFiles A successful 1`] = `
{
  "logs": [
    "injected in src/strings.ts : 7 times",
  ],
  "totalInjections": 7,
}
`;

exports[`injectMarkInFiles B fail to find placeholder 1`] = `
"could not find a place to inject in src/strings.ts, aborting.

Please use one or more of these placeholders :  <span id=\\"nope\\"></span>  <meta name=\\"nope\\" content=\\"\\">  __nope__"
`;
